Assets/Scripts/Controller/WaypointController.cs

```diff
@@ -1,22 +1,27 @@
-using System;
 using UnityEngine;
 
 // The waypoint controller steers the agent to the target using direct linear guidance.
-public class WaypointController : IControllerLegacy {
-  // Desired speed in m/s.
-  float _desiredSpeed;
+public class WaypointController : ControllerBase {
+  // To prevent overshooting near the waypoint, if the squared distance to the waypoint is less than
+  // the threshold, do not apply any acceleration.
+  private const float _accelerationCutoffDistanceSqr = 1.0f;
 
-  public WaypointController(Agent agent, float desiredSpeed) : base(agent) {
-    _desiredSpeed = desiredSpeed;
-  }
+  public WaypointController(IAgent agent) : base(agent) {}
+
+  // Controller-dependent implementation of the control law.
+  protected override Vector3 Plan(in Transformation relativeTransformation) {
+    Vector3 relativePosition = relativeTransformation.Position.Cartesian;
+    if (relativeTransformation.Position.Range < _accelerationCutoffDistanceSqr) {
+      return Vector3.zero;
+    }
 
-  protected override Vector3 PlanImpl(in Transformation relativeTransformation) {
-    Vector3 toWaypoint = relativeTransformation.Position.Cartesian;
-    Vector3 desiredVelocity = toWaypoint.normalized * _desiredSpeed;
+    // To reach the waypoint as fast as possible, use the maximum forward acceleration.
+    Vector3 forward = Agent.transform.forward;
+    Vector3 forwardAccelerationInput = forward * Agent.MaxForwardAcceleration();
 
-    // Calculate the acceleration needed to reach the desired velocity within one time step.
-    Vector3 accelerationInput =
-        (desiredVelocity - _agent.GetVelocity()) / (float)Time.fixedDeltaTime;
-    return accelerationInput;
+    // Steer as hard as possible towards the waypoint.
+    Vector3 normalAcceleration = Vector3.ProjectOnPlane(relativePosition, forward).normalized;
+    Vector3 normalAccelerationInput = normalAcceleration * Agent.MaxNormalAcceleration();
```

Comment:
nit: NRE
The `WaypointController` calls `Agent.MaxForwardAcceleration()` and `Agent.MaxNormalAcceleration()` without ensuring `Agent.StaticConfig` is not null. While `AgentBase.MaxForwardAcceleration()` has null-coalescing (`??`), the test setup in WaypointControllerTests.cs manually sets `StaticConfig` - this is a dependency that's not obvious.